internal/cmdrender/runtime/container.go

```diff
@@ -41,6 +44,34 @@ type ContainerFnPermission struct {
 	AllowNetwork bool
 }
 
+// ContainerFnWrapper wraps the real function filter, prints
+// the function running progress and failures.
+type ContainerFnWrapper struct {
+	Fn *ContainerFn
+}
+
+func (fw *ContainerFnWrapper) Run(r io.Reader, w io.Writer) error {
+	pr := printer.FromContextOrDie(fw.Fn.Ctx)
+	printOpt := printer.NewOpt().Indent(printer.FnIndentation)
+	pr.OptPrintf(printOpt, "[RUNNING] %q\n", fw.Fn.Image)
+	err := fw.Fn.Run(r, w)
+	if err != nil {
+		pr.OptPrintf(printOpt, "[FAIL] %q\n", fw.Fn.Image)
+		failureOpt := printer.NewOpt().
+			Indent(fnerrors.FnExecErrorIndentation).
+			Stderr()
+		fe, ok := err.(*fnerrors.FnExecError)
```

Comment:
tip: use `errors.As()` as much as you can.

```
var fnErr FnExecError

if !errors.As(err, &fnErr) {
  return err
}
...
return fnErr // return an instance of fnErr here so that errorResolver for fnExecError type can help resolve it. That will suppress the default error printing.

```

Reply:
done